src/ucs/datastruct/usage_tracker.h

```diff
@@ -0,0 +1,180 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2023. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_USAGE_TRACKER_H_
+#define UCS_USAGE_TRACKER_H_
+
+
+#include <stdint.h>
+#include <stddef.h>
+
+
+#include <ucs/datastruct/khash.h>
+#include <ucs/datastruct/lru.h>
+
+
+/* Usage Tracker element data structure */
+typedef struct {
+    /* Key to use as hash table input */
+    void  *key;
+    /* Exponential decay based score ([0-1]) */
+    double score;
+    /* Lower bound to score value */
+    double min_score;
+    /* Is this entry considered 'promoted' */
+    int    promoted;
+} ucs_usage_tracker_element_t;
+
+
+/* Callback type for rank modify notification */
+typedef void (*ucs_usage_tracker_rank_cb_t)(void *entry, void *arg);
+
+
+/* Usage Tracker exponential decay parameters */
+typedef struct {
+    /* Multiplier factor [0-1] */
+    double a;
+    /* Addition factor [0-1] */
+    double b;
+} ucs_usage_tracker_exp_decay_t;
+
+
+typedef struct {
+    /* Max number of promoted entries */
+    unsigned                    promote_capacity;
+    /* Max number of entries to promote in each progress */
+    unsigned                    promote_thresh;
+    /* Min score difference in order to remove an entry from promoted list [0-1]*/
+    double                      remove_thresh;
+    /* User callback which will be called when an entry is promoted. */
+    ucs_usage_tracker_rank_cb_t promote_cb;
+    /* User object which will be passed to promote callback. */
+    void                       *promote_arg;
+    /* User callback which will be called when an entry is demoted. */
+    ucs_usage_tracker_rank_cb_t demote_cb;
+    /* User object which will be passed to demote callback. */
+    void                       *demote_arg;
+    /* Exponential decay linear parameters (mult/add factors) */
+    ucs_usage_tracker_exp_decay_t exp_decay;
+} ucs_usage_tracker_params_t;
+
+
+/* Hash table type for Usage Tracker class */
+KHASH_INIT(usage_tracker_hash, uint64_t, ucs_usage_tracker_element_t, 1,
+           kh_int64_hash_func, kh_int64_hash_equal);
+
+

```

Comment:
remove these spaces

Reply:
you mean typedef should be in the next line?